src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -1122,15 +1198,45 @@ private void PerformSyncOperation<TOperation>(ref OperationQueue<TOperation> que
                     return;
                 }
 
-                if (e.Wait(timeout))
+                bool timeoutExpired = false;
+                while (true)
                 {
-                    // Completed within timeout
-                    return;
+                    DateTime waitStart = DateTime.UtcNow;
+
+                    if (!e.Wait(timeout))
+                    {
+                        timeoutExpired = true;
+                        break;
+                    }
+
+                    // Reset the event now to avoid lost notifications if the processing is unsuccessful.
+                    e.Reset();
```

Comment:
I'm not understanding this part.  What happens if the operation completes and the event is set just after Wait returns false?  Will that then be handled by the subsequent check for Completed/Canceled?

Reply:
The event won't be set in that case.

One of the following will happen:
(1) The operation completes, is removed from the queue, and next op (if any) is dispatched.  In this case the event is gone from the queue and will never be set again.
(2) The operation doesn't complete and the queue goes back to Waiting state.  As soon as it becomes Waiting, it can transition back to Processing (due to another readiness notification coming through), which will cause us to set the event.  So we can't reset the event after it transitions back to Waiting; we need to do it before. 

I could change this to do the Reset only in the case where we actually transition to Waiting, but not sure it's worth it.